repo URL: https://github.com/<redacted>/assignment-4-equation-of-a-slime-<redacted>\
commitID: 548d1a5ce66b40615bf44ceb8fcb649b001c637f

In [19]:
# Imports section
import pandas as pd
import sklearn as sk
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import Pipeline

## Part 1. Loading the dataset

In [20]:
# Using pandas load the dataset (load remotely, not locally)
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)
df = pd.read_csv('./science_data_large.csv')
print(df.head(15))
df.info()

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## Part 2. Splitting the dataset

In [21]:
# Take the pandas dataset and split it into our features (X) and label (y)
dfCopy = df.copy()
x = dfCopy.drop('Size nm^3', axis = 1)
y = dfCopy.drop(['Temperature °C', 'Mols KCL'], axis = 1)
# print(x.head(10))
# print(y.head(10))

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
xTrain, xTest, yTrain, yTest = sk.model_selection.train_test_split(x, y,  test_size = 0.1)

## Part 3. Perform a Linear Regression

In [22]:
# Use sklearn to train a model on the training set
linRegModel = LinearRegression()
linRegModel.fit(xTrain, yTrain)

# Create a sample datapoint and predict the output of that sample with the trained model
sample = pd.DataFrame({'Temperature °C':[660], 'Mols KCL':[519]})
prediction = (linRegModel.predict(sample)) 
print(f'datapoint:\n{sample};\npredicted output {prediction[0][0]:.5f}') 

# Report on the score for that model, in your own words (markdown, not code) explain what the score means
score = linRegModel.score(xTest, yTest)
print(f'score for the model is {score:.5f}') 

# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
params = linRegModel.coef_
intercept = linRegModel.intercept_
print(f'params: {params[0][0]:.5f},{params[0][1]:.5f}\nintercept: {(intercept[0]):.5f}')



datapoint:
   Temperature °C  Mols KCL
0             660       519;
predicted output 697076.38455
score for the model is 0.87094
params: 880.16273,1029.19219
intercept: -417981.76324


The score means that the model is correctly predicting the data around 88% of the time. 

$$h(x) = -417981.76324 + 880.16273x_1 + 1029.19219x_2$$

Sample equation: $E = mc^2$

## Part 4. Use Cross Validation

In [23]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
cvScore = sk.model_selection.cross_val_score(linRegModel, x, y) 
print(cvScore)

# Report on their finding and their significance
print(f'mean: {cvScore.mean():.5f}\nstandard deviation: {cvScore.std():.5f}')

[0.83918826 0.87051239 0.85871066 0.87202623 0.84364641]
mean: 0.85682
standard deviation: 0.01347


By segmenting the data into five partitions, with each partition being seen as the test data once, the model has around 86% acurracy with standard deviation of 0.013 in predicting labels. This is about 2% less accurate than the score of just 90% training and 10% testing. Although the accuracy might be less, we do this to ensure that the model isn't overfitting the data and works for arbitrary values.

## Part 5. Using Polynomial Regression

In [24]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
polyModel = Pipeline([('poly', PolynomialFeatures(degree=2)), ('linear', LinearRegression(fit_intercept=False))])
polyModel.fit(xTrain, yTrain) 

# Report on the metrics and output the resultant equation as you did in Part 3.
polyScore = polyModel.score(xTest, yTest)
print(f'score for the model is {polyScore:.5f}') 

linMod = polyModel.named_steps['linear']
polyParams = linMod.coef_
polyIntercept = linMod.intercept_
print(f'params: {polyParams}\nintercept: {(polyIntercept)}')

score for the model is 1.00000
params: [[ 1.41683162e-05  1.20000000e+01 -1.12782249e-07 -1.96180849e-11
   2.00000000e+00  2.85714287e-02]]
intercept: 0.0


The score means that the model is predicting data accurately 100% of the time. 

$$h(x) = 12x_1 + 2x_1x_2 + 0.02857x_{2}^2$$

*Other terms evaluate to 0 and thus are left out.